runtime/wasm/src/module/mod.rs

```diff
@@ -269,126 +123,367 @@ impl WasmiModule {
+        link!("bigDecimal.minus", big_decimal_minus, x_ptr, y_ptr);
+        link!("bigDecimal.times", big_decimal_times, x_ptr, y_ptr);
+        link!("bigDecimal.divided_by", big_decimal_divided_by, x, y);
+        link!("bigDecimal.equals", big_decimal_equals, x_ptr, y_ptr);
+
+        link!("dataSource.create", data_source_create, name, params);
+        link!(
+            "dataSource.createWithContext",
+            data_source_create_with_context,
+            name,
+            params,
+            context
         );
+        link!("dataSource.address", data_source_address,);
+        link!("dataSource.network", data_source_network,);
+        link!("dataSource.context", data_source_context,);
+
+        link!("ens.nameByHash", ens_name_by_hash, ptr);
+
+        link!("log.log", log_log, level, msg_ptr);
+
+        link!("arweave.transactionData", arweave_transaction_data, ptr);
+
+        link!("box.profile", box_profile, ptr);
+
+        let instance = linker.instantiate(&valid_module.module)?;
+
+        // Provide access to the WASM runtime linear memory
+        let memory = instance
+            .get_memory("memory")
+            .context("Failed to find memory export in the WASM module")?;
+
+        let memory_allocate = instance
+            .get_func("memory.allocate")
+            .context("`memory.allocate` function not found")?
+            .get1()?;
+
+        let should_interrupt = Arc::new(AtomicBool::new(true));
+        if let Some(timeout) = timeout {
+            // This task is likely to outlive the instance, which is fine.
+            let interrupt_handle = instance.store().interrupt_handle().unwrap();
+            let should_interrupt = should_interrupt.clone();
+            graph::spawn(async move {
+                tokio::time::delay_for(timeout).await;
+                if should_interrupt.load(Ordering::SeqCst) {
+                    interrupt_handle.interrupt()
+                }
+            });
+        }
+
+        *shared_instance.borrow_mut() = Some(WasmInstance {
+            instance,
+            memory_allocate: Box::new(memory_allocate),
+            memory,
+            ctx,
+            valid_module,
+            host_metrics,
+            timeout,
+            should_interrupt,
+
+            // `arena_start_ptr` will be set on the first call to `raw_new`.
+            arena_free_size: 0,
+            arena_start_ptr: 0,
+        });
 
-        result.map(|_| self.ctx.state).map_err(|err| {
-            format_err!(
-                "Failed to handle Ethereum block with handler \"{}\": {}",
-                handler_name,
-                format_wasmi_error(err)
-            )
+        Ok(WasmInstanceHandle {
+            instance: shared_instance.clone(),
         })
     }
 }
 
-impl AscHeap for WasmiModule {
-    fn raw_new(&mut self, bytes: &[u8]) -> Result<u32, Error> {
+impl AscHeap for WasmInstance {
+    fn raw_new(&mut self, bytes: &[u8]) -> u32 {
```

Comment:
Nope, we even use an AS allocator that doesn't have `free`. The reason is that we have no idea of when it would be ok to free the objects we allocate here. The latest AS versions have a GC, and maybe it would be interesting to turn that on for the objects allocated within the mapping, but generally the lack of `free` hasn't been a problem.